### PR TITLE
translate tempo config to pydantic models

### PR DESCRIPTION
**Pull request by michaeldmitry**
## Issue
tempo configurations are in the form of raw dicts.

## Solution
Translate all currently used tempo config to pydantic models.

## Testing instructions
For coordinator to work and push tempo configuration to workers, it needs an s3 integration. To workaround the issue and not need an actual s3 instance, we can use Facade charm for mocking s3 interface.

### Deploy tempo k8s operator (used for validation)
```
juju deploy tempo-k8s tempo --resource tempo-image=grafana/tempo:2.4.0 --channel=latest/edge
```

### Deploy tempo coordinator
cwd to `tempo-coordinator-k8s-operator`

```
charmcraft pack
juju deploy ./tempo-coordinator-k8s_ubuntu-22.04-amd64.charm 
```

### Deploy Facade
cwd to `tempo-coordinator-k8s-operator`
```
juju deploy facade s3-facade --channel=latest/edge
juju integrate s3-facade:provide-s3 tempo-coordinator-k8s:s3
juju run s3-facade/0 update --params ./tests/manual/facades/s3.yaml
```

### Deploy tempo worker
cwd to `tempo-worker-k8s-operator`

```
git fetch
git checkout coordinator-integration
charmcraft pack
juju deploy ./tempo-worker-k8s_ubuntu-22.04-amd64.charm  tempo-worker --resource tempo-image=docker.io/ubuntu/tempo:2-22.04
```
### Integrate cluster
```
 juju integrate tempo-coordinator-k8s tempo-worker
```

### Validation

1- coordinator and worker should both be in active/idle
2- kubectl exec -it tempo-worker-0 -n <model> -c tempo -- cat /etc/tempo/tempo.yaml
3- kubectl exec -it tempo-0 -n <model> -c tempo -- cat /etc/tempo/tempo.yaml
4- Both config files should match apart from storage options, as s3 would be used in coordinator instead of local storage

### TLS Validation
Repeat above steps after ` juju integrate ssc tempo-coordinator-k8s`

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -466,6 +466,9 @@ def _update_tempo_cluster(self):
             logger.error("skipped tempo cluster update: inconsistent state")
             return
 
+        if not self.unit.is_leader():
+            return
```
Comment (sed-i) [inline src/charm.py:470]: Since you have a leader guard, do we need to make sure that the code below is run on `leader-elected`?

File: src/tempo.py [hunk 2]
```diff
@@ -132,21 +151,21 @@ def get_receiver_url(self, protocol: ReceiverProtocol, ingress: TraefikRouteRequ
...
+
         if self.use_tls:
-            for cfg in ("http_tls_config", "grpc_tls_config"):
-                server_config[cfg] = {  # type: ignore
-                    "cert_file": str(self.tls_cert_path),
-                    "key_file": str(self.tls_key_path),
-                    "client_ca_file": str(self.tls_ca_path),
-                    "client_auth_type": "VerifyClientCertIfGiven",
-                }
-            server_config["tls_min_version"] = self._tls_min_version  # type: ignore
```
Comment (sed-i) [inline src/tempo.py:149]: Should we also remove the `_tls_min_version = ""` class attr?

File: src/tempo_cluster.py [hunk 3]
```diff
@@ -161,10 +163,10 @@ def dump(self, databag: Optional[MutableMapping] = None, clear: bool = True):
                 databag = {}
 
             if self._NEST_UNDER:
-                databag[self._NEST_UNDER] = self.json(by_alias=True, exclude_defaults=True)
+                databag[self._NEST_UNDER] = self.json(by_alias=True, exclude_none=True)
```
Comment (sed-i) [inline src/tempo_cluster.py:166]: Do we still need `_NEST_UNDER`?
Iirc, it was introduced into traefik because at the time the old pydantic had some limitations.

File: src/tempo_config.py
Comment (sed-i) [inline src/tempo_config.py:1]: It's great that the config is ops independent. Let's try to keep it that way.

Re pydantic model itself (and how strict it is in loading rel data): how would an upgrade sequence look like when the new version has a schema change? Could we get tracebacks for validation error between coordinator and worker?

File: tests/integration/test_ingressed_tls.py
Comment (sed-i) [inline tests/integration/test_ingressed_tls.py:1]: Should we use a pytest "skip" call instead of commenting out the entire file?
Not sure which approach would be more noticeable down the road.

File: src/charm.py (hunk 1 above)
Comment (michaeldmitry) [reply src/charm.py:470]: Yes. It's already run on `leader-elected`

File: tests/integration/test_ingressed_tls.py
Comment (michaeldmitry) [reply tests/integration/test_ingressed_tls.py:1]: I think in terms of "noticeability", commenting the file is more visible. Anyways, its not going to hang like that for long. Should be uncommented, then fixed once we have a fully working cluster.

File: src/tempo_config.py
Comment (michaeldmitry) [reply src/tempo_config.py:1]: After discussion with @sed-i 
relation data for tempo config is sent as a dictionary, only the construction of the config in coordinator is done through pydantic models. And this dictionary is used by the dumb workers as is to run the workload.
So, in the event of changing the schema of the tempo configuration with a breaking change, workers would fail with an error in running the workload (or something similar). 
A suggestion was to try to validate the received dict on the model in workload side and if it fails, we would put the worker in Blocking state instead of crashing with an error.

@PietroPasotti  @mmkay what are your thoughts on this matter?

Comment (michaeldmitry) [reply src/tempo_config.py:1]: I think we don't even need to validate against the pydantic model.
https://github.com/grafana/tempo/issues/2131
Tempo itself can help us verify if the passed config file is correct or not.

**Review by mmkay — APPROVED** [review]

File: src/tempo_cluster.py [hunk 4]
```diff
@@ -180,18 +174,12 @@ class DatabagModel(BaseModel):
...
-            # Custom config key: whether to nest the whole datastructure (as json)
-            # under a field or spread it out at the toplevel.
-            _NEST_UNDER=None,
         )  # type: ignore
         """Pydantic config."""
 
         @classmethod
         def load(cls, databag: MutableMapping):
             """Load this model from a Juju databag."""
-            nest_under = cls.model_config.get("_NEST_UNDER")
```
Comment (PietroPasotti) [inline src/tempo_cluster.py:192]: uhm, why did you remove the nest_under logic from the whole object?

File: src/tempo.py [hunk 5]
```diff
@@ -132,21 +147,21 @@ def get_receiver_url(self, protocol: ReceiverProtocol, ingress: TraefikRouteRequ
         return f"{url}:{receiver_port}"
 
     def _build_server_config(self):
-        server_config = {
-            "http_listen_port": self.tempo_http_server_port,
+        server_config = Server(
```
Comment (PietroPasotti) [inline src/tempo.py:150]: all other types have a Config suffix, I think this should be ServerConfig as well.

File: src/tempo.py [hunk 6]
```diff
@@ -212,42 +197,34 @@ def generate_config(
...
+            backend="s3",
+            s3=S3Config(
+                bucket=s3_config["bucket"],
+                access_key=s3_config["access-key"],
+                endpoint=s3_config["endpoint"],
+                secret_key=s3_config["secret-key"],
+            ),
+            block=BlockConfig(version="v2"),
+        )
+        return Storage(trace=storage_config)
```
Comment (PietroPasotti) [inline src/tempo.py:227]:
```suggestion
        return StorageConfig(trace=storage_config)
```

File: src/tempo.py [hunk 7]
```diff
@@ -265,7 +242,51 @@ def is_ready(self):
             return False
         return out == "ready"
 
-    def _build_receivers_config(self, receivers: Sequence[ReceiverProtocol]):  # noqa: C901
+    def _build_querier_config(self):
+        """Build querier config"""
+        # TODO this won't work for distributed coordinator where query frontend will be on a different unit
+        return Querier(
```
Comment (PietroPasotti) [inline src/tempo.py:248]:
```suggestion
        return QuerierConfig(
```

File: src/tempo.py [hunk 8]
```diff
@@ -265,7 +242,51 @@ def is_ready(self):
...
+            frontend_worker=FrontendWorker(
```
Comment (PietroPasotti) [inline src/tempo.py:249]:
```suggestion
            frontend_worker=FrontendWorkerConfig(
```

or, if you don't like the idea of having 100 objects all called SomethingConfig, I'd be fine with doing:
```
from xx import tempo_config
tempo_config.Querier
tempo_config.Server
```
and so on. Then the prefix makes it clear we're talking about the configuration of a server, and not the server itself.

File: src/tempo_cluster.py [hunk 9]
```diff
@@ -22,6 +22,8 @@
 from ops import EventSource, Object, ObjectEvents
 from pydantic import BaseModel, ConfigDict
 
+from tempo_config import TempoBaseConfig
```
Comment (PietroPasotti) [inline src/tempo_cluster.py:25]: I'd rename this to `TempoConfig`. The name 'base' usually indicates you're meant to subclass it or specialize it somehow.

File: tests/integration/test_ingressed_tls.py [hunk 10]
```diff
@@ -1,180 +1,177 @@
...
-async def test_remove_relation(ops_test: OpsTest):
-    await ops_test.juju(
-        "remove-relation", APP_NAME + ":certificates", SSC_APP_NAME + ":certificates"
-    )
-    await asyncio.gather(
-        ops_test.model.wait_for_idle(
-            apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=1000
-        ),
-    )
+# import asyncio
```
Comment (PietroPasotti) [inline tests/integration/test_ingressed_tls.py:1]: why are we commenting out a whole module? 
If it's no longer relevant, let's remove it. 
If there's a really good reason to keep it around, write that in a FIXME/TODO.

File: tests/integration/test_charm.py [hunk 11]
```diff
@@ -33,9 +33,6 @@ async def test_build_and_deploy(ops_test: OpsTest):
           {mc.name}:
             charm: {charm}
             trust: true
-            resources:
```
Comment (PietroPasotti) [inline tests/integration/test_charm.py:36]: were these changes necessary?

File: tests/integration/test_integration.py [hunk 12]
```diff
@@ -1,168 +1,165 @@
...
-        ops_test.model.wait_for_idle(
-            apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=1000
-        ),
-        # for tester, depending on the result of race with tempo it's either waiting or active
-        ops_test.model.wait_for_idle(apps=[TESTER_APP_NAME], raise_on_blocked=True, timeout=1000),
-        ops_test.model.wait_for_idle(
-            apps=[TESTER_GRPC_APP_NAME], raise_on_blocked=True, timeout=1000
-        ),
-    )
+# import asyncio
```
Comment (PietroPasotti) [inline tests/integration/test_integration.py:1]: same

File: tests/integration/test_scaling_monolithic.py [hunk 13]
```diff
@@ -12,9 +13,10 @@
 from pytest_operator.plugin import OpsTest
 
 METADATA = yaml.safe_load(Path("./charmcraft.yaml").read_text())
-APP_NAME = "tempo"
+APP_NAME = "tempo-coordinator"
```
Comment (PietroPasotti) [inline tests/integration/test_scaling_monolithic.py:16]: mimir's coordinator is also deployed as 'mimir' in the bundle. I'd not change this.

File: tests/integration/test_scaling_monolithic.py [hunk 14]
```diff
@@ -65,19 +67,35 @@ def present_facade(
...
-    with tempfile.NamedTemporaryFile() as f:
+    with tempfile.NamedTemporaryFile(dir=os.getcwd()) as f:
         fpath = Path(f.name)
         fpath.write_text(yaml.safe_dump(data))
 
         _model = f" --model {model}" if model else ""
-        run(shlex.split(f"juju run {app}/0{_model} --params {fpath.absolute()}"))
+
+        run(shlex.split(f"juju run {app}/0{_model} update --params {fpath.absolute()}"))
+        # facade charm edge rev9 copies data into 'mocks/provide' not 'mocks/require'
```
Comment (PietroPasotti) [inline tests/integration/test_scaling_monolithic.py:77]: this should be fixed now.

File: tests/integration/test_scaling_monolithic.py [hunk 15]
```diff
@@ -23,13 +25,13 @@
...
-    await ops_test.model.deploy(tempo_charm, resources=resources, application_name=APP_NAME),
+    await ops_test.model.deploy(tempo_charm, application_name=APP_NAME)
 
     await ops_test.model.wait_for_idle(
         apps=[APP_NAME],
-        # tempo might be in waiting as it waits for tempo workload ready
-        raise_on_blocked=True,
+        # coordinator will be blocked on s3 and workers integration
+        status="blocked",
+        raise_on_blocked=False,
```
Comment (PietroPasotti) [inline tests/integration/test_scaling_monolithic.py:34]: I _think_ if you expect status="blocked" it should also implicitly not raise on blocked.

File: tests/integration/test_scaling_monolithic.py [hunk 16]
```diff
@@ -65,19 +67,35 @@ def present_facade(
...
 
         _model = f" --model {model}" if model else ""
-        run(shlex.split(f"juju run {app}/0{_model} --params {fpath.absolute()}"))
+
+        run(shlex.split(f"juju run {app}/0{_model} update --params {fpath.absolute()}"))
+        # facade charm edge rev9 copies data into 'mocks/provide' not 'mocks/require'
+        # workaround to mv the copied file to the correct path inside 'require' directory
+        # until charm-relation-interfaces/pull/152 is merged.
+        if role == "require":
+            run(
```
Comment (PietroPasotti) [inline tests/integration/test_scaling_monolithic.py:81]: oh man so sorry you had to do this horrible thing :D

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
impressive work! 
I have a few suggestions about the API (class names, etc...) that are non-blocking
as for the commented-out modules, let's add an explanation of what's going on or let's remove them.

File: src/tempo.py [hunk 17]
```diff
@@ -212,42 +197,34 @@ def generate_config(
...
-                "secret_key": s3_config["secret-key"],
-                "insecure": (False if s3_config["endpoint"].startswith("https://") else True),
-            },
-        }
-        return {"trace": storage_config}
+                max_workers=400,
+                queue_depth=20000,
+            ),
+            backend="s3",
+            s3=S3Config(
```
[the remark this answers is not in the document]
Comment (michaeldmitry) [reply src/tempo.py:219]: I don't think it would make much of a difference if you're referring to the same level of validation we currently do (keys exist with correct value types). 
once we get the `credentials_changed` event, we generate the config (`_update_tempo_cluster`). 
Also, I think it'd be cleaner to keep construction of config with the validation it offers in one place.

File: src/tempo_cluster.py (hunk 4 above)
Comment (michaeldmitry) [reply src/tempo_cluster.py:192]: @sed-i raised a comment if we still need that since it might have been introduced due to old pydantic limitation.
Seems that with the pydantic version that gets used in this charm, things work normally without this object and its logic.

File: tests/integration/test_charm.py (hunk 11 above)
Comment (michaeldmitry) [reply tests/integration/test_charm.py:36]: Yes. resources "key" doesn't exist in this current charmcraft. it breaks the test.

File: src/tempo_config.py
Comment (PietroPasotti) [reply src/tempo_config.py:1]: So my thought here is we should try to fail early and loudly. 
**Option A: validate once.**
The coordinator generates a config, the pydantic model validates that it's correct. We put it in relation data.
Worker picks it up, puts it in the workload, workload breaks because the config is not valid.
Now we have to figure out that the workload is broken and why. 

**Option B: validate twice.**
The coordinator generates a config, the pydantic model validates that it's correct. We put it in relation data.
Worker picks it up, attempts to validate it but it fails. Worker immediately sets blocked status and goes belly up, saying 'config invalid'.

Regardless of _how_ we do the validation in the worker (whether it's by the built-in [cli tool offered by tempo](https://github.com/grafana/tempo/issues/2131) or by a copy of the same pydantic model hosted in the controller, that's up to us) I think we should.

**Option C: validate thrice.**
The coordinator generates a config, the pydantic model validates that it's correct. We put it in relation data.
Worker picks it up, attempts to validate it with (the same) pydantic model and it succeeds. 
It attempts to validate it with the tempo cli tool and it fails. 
This means that the pydantic model and the tempo image are out of sync: our models encode a config that's not supported in the tempo version we're running. 

Option C feels a bit like overkill, we can guard against this scenario by integration-testing the worker and verify that the config generated by the pydantic models always passes the built-in config validator. So I'd say let's go for B.
As to how we validate, I think Pydantic models offers us the most flexibility, because in case we want to be backwards-compatible or support consecutive tempo versions, the worker could implement logic like so:
```
if ConfigModelV1.validate(config_from_reldata):
    unit.status = ActiveStatus("Please upgrade the tempo coordinator to v2 (rev > 1412)")
   conf = ConfigModelV2.from_v1(config_from_reldata) 
else:
    conf = ConfigModelV2(config_from_reldata)
```

and the upgrade path would be:
- upgrade the workers
- upgrade the coordinator

File: src/tempo.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply src/tempo.py]: I do like the idea to keep the validation in one place

File: tests/integration/test_ingressed_tls.py (hunk 10 above)
Comment (PietroPasotti) [reply tests/integration/test_ingressed_tls.py:2]: how was this issue resolved exactly?

File: src/tempo_config.py [hunk 18]
```diff
@@ -24,143 +22,184 @@ class InvalidConfigurationError(Exception):
...
 
-    ingestion_rate: int = 0
-    ingestion_burst_size: int = 0
-    max_global_series_per_user: int = 0
-    ruler_max_rules_per_rule_group: int = 0
-    ruler_max_rule_groups_per_tenant: int = 0
+    abort_if_cluster_join_fails: bool
+    bind_port: int
+    join_members: List[str]
+    tls_enabled: Optional[bool] = False
```
Comment (PietroPasotti) [inline src/tempo_config.py:44]: is this truly optional? as in, can it be None? or is it really bool (given that the default is False...)

File: src/tempo_config.py [hunk 19]
```diff
@@ -24,143 +22,184 @@ class InvalidConfigurationError(Exception):
...
-    """Alertmanager schema."""
+class TLS(BaseModel):
+    """TLS configuration schema."""
 
-    data_dir: str = "/data/alertmanager"
-    external_url: Optional[str]
+    cert_file: str
+    key_file: str
+    client_ca_file: str
+    client_auth_type: Optional[str] = "VerifyClientCertIfGiven"
```
Comment (PietroPasotti) [inline src/tempo_config.py:102]: this feels like it should be an enum

File: src/tempo_config.py [hunk 20]
```diff
@@ -24,143 +22,184 @@ class InvalidConfigurationError(Exception):
...
+    """Pool schema."""
+
+    max_workers: int
+    queue_depth: int
+
+
+class Wal(BaseModel):
+    """Wal schema."""
+
+    path: str
```
Comment (PietroPasotti) [inline src/tempo_config.py:141]: I think pydantic takes `pathlib.Path` as well? perhaps that gives better validation?

File: src/tempo_config.py [hunk 21]
```diff
@@ -24,143 +22,184 @@ class InvalidConfigurationError(Exception):
...
+
+    bucket: str
+    access_key: str
     endpoint: str
-    secret_access_key: str = Field(alias="secret-key")
-    bucket_name: str = Field(alias="bucket")
-    region: str = ""
-    insecure: str = "false"
+    secret_key: str
+    insecure: Optional[bool] = False
```
Comment (PietroPasotti) [inline src/tempo_config.py:151]: just bool?

**Review by PietroPasotti — COMMENTED** [review]
a few possible typing issues with the models

File: tests/integration/test_ingressed_tls.py (hunk 10 above)
Comment (michaeldmitry) [reply tests/integration/test_ingressed_tls.py:2]: Added a TODO

File: src/tempo_config.py (hunk 18 above)
Comment (michaeldmitry) [reply src/tempo_config.py:44]: its semantically misleading. Will make it a bool

File: src/tempo_config.py [hunk 22]
```diff
@@ -3,164 +3,217 @@
...
 S3_RELATION_NAME = "s3"
 BUCKET_NAME = "tempo"
 
 logger = logging.getLogger(__name__)
 
 
-class InvalidConfigurationError(Exception):
-    """Invalid configuration."""
+class ClientAuthTypeEnum(str, enum.Enum):
+    """client auth types"""
```
Comment (PietroPasotti) [inline src/tempo_config.py:22]:
```suggestion
    """Client auth types."""
```

File: src/tempo_config.py [hunk 23]
```diff
@@ -3,164 +3,217 @@
...
 
-    pass
+    VERIFY_CLIENT_CERT_IF_GIVEN = "VerifyClientCertIfGiven"
```
Comment (PietroPasotti) [inline src/tempo_config.py:24]: please add the other possible values:
https://pkg.go.dev/crypto/tls#ClientAuthType

**Review by PietroPasotti — APPROVED** [review]
once the last couple of issues with the enum are fixed, I'm fine with merging :) 
Thanks!

File: src/tempo_config.py (hunk 23 above)
Comment (PietroPasotti) [reply src/tempo_config.py:24]: (and maybe add this link in a comment to document where we're grabbing these values from)
